In [1]:
import pandas as pd
import numpy as np

In [2]:
fdata = pd.read_csv('flight_data.csv')
fdata.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [3]:
fdata.dtypes

year                int64
month               int64
day                 int64
dep_time          float64
sched_dep_time      int64
dep_delay         float64
arr_time          float64
sched_arr_time      int64
arr_delay         float64
carrier            object
flight              int64
tailnum            object
origin             object
dest               object
air_time          float64
distance            int64
hour                int64
minute              int64
time_hour          object
dtype: object

# Check Null Values in Dataset  : 



In [4]:
fdata.isnull().values.any()

True

In [5]:
fdata.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [6]:
len(fdata[fdata['arr_time'].isnull() & fdata['dep_time'].isnull()])

8255

* There are 8255 rows in which both arrival & departure time is not available. We can drop above rows from dataset! 

In [7]:
fdata.shape

(336776, 19)

In [8]:

fdata.dropna(subset=['dep_time', 'arr_time'], how='all',inplace=True)

In [9]:
fdata.shape


(328521, 19)

In [10]:
fdata.isnull().sum()

year                 0
month                0
day                  0
dep_time             0
sched_dep_time       0
dep_delay            0
arr_time           458
sched_arr_time       0
arr_delay         1175
carrier              0
flight               0
tailnum              0
origin               0
dest                 0
air_time          1175
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [11]:
fdata['arr_time'].isnull().sum()

458

* There are 458 Flights for which arrival time is NOT available but departure time is available. These are valid number of flights where dep_time data is missing. -- Keep these rows


* We can fill these rows by adding tavel time in dep_time . sched_arr_time - sched_dep_time

In [12]:
fdata["dep_time"]= pd.to_numeric(fdata['dep_time']).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["sched_dep_time"]= pd.to_numeric(fdata['sched_dep_time']).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["sched_arr_time"]= pd.to_numeric(fdata['sched_arr_time']).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["dep_delay"]= pd.to_numeric(fdata['dep_delay']).astype(np.int64)

In [13]:
fdata.head()

year  month  day dep_time sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     0517           0515          2     830.0   
1  2013      1    1     0533           0529          4     850.0   
2  2013      1    1     0542           0540          2     923.0   
3  2013      1    1     0544           0545         -1    1004.0   
4  2013      1    1     0554           0600         -6     812.0   

  sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0           0819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1           0830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2           0850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3           1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4           0837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [14]:
fdata["dep_time"]= pd.to_timedelta(fdata["dep_time"].str[:2]+":"+fdata["dep_time"].str[-2:]+":00")
fdata["sched_dep_time"]= pd.to_timedelta(fdata["sched_dep_time"].str[:2]+":"+fdata["sched_dep_time"].str[-2:]+":00")
fdata["sched_arr_time"]= pd.to_timedelta(fdata["sched_arr_time"].str[:2]+":"+fdata["sched_arr_time"].str[-2:]+":00")

In [15]:
fdata.head()

year  month  day dep_time sched_dep_time  dep_delay  arr_time  \
0  2013      1    1 05:17:00       05:15:00          2     830.0   
1  2013      1    1 05:33:00       05:29:00          4     850.0   
2  2013      1    1 05:42:00       05:40:00          2     923.0   
3  2013      1    1 05:44:00       05:45:00         -1    1004.0   
4  2013      1    1 05:54:00       06:00:00         -6     812.0   

  sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0       08:19:00       11.0      UA    1545  N14228    EWR  IAH     227.0   
1       08:30:00       20.0      UA    1714  N24211    LGA  IAH     227.0   
2       08:50:00       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3       10:22:00      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4       08:37:00      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [16]:
fdata[fdata['arr_time'].isnull() == True]

year  month  day dep_time sched_dep_time  dep_delay  arr_time  \
754     2013      1    1 20:16:00       19:30:00         46       NaN   
1714    2013      1    2 20:41:00       20:45:00         -4       NaN   
1756    2013      1    2 21:45:00       21:29:00         16       NaN   
7039    2013      1    9 06:15:00       06:15:00          0       NaN   
7851    2013      1    9 20:42:00       20:40:00          2       NaN   
9284    2013      1   11 13:44:00       13:50:00         -6       NaN   
11128   2013      1   13 19:07:00       16:34:00        153       NaN   
11245   2013      1   13 22:39:00       21:59:00         40       NaN   
13268   2013      1   16 08:37:00       08:40:00         -3       NaN   
21425   2013      1   25 14:52:00       15:00:00         -8       NaN   
21435   2013      1   25 14:57:00       15:05:00         -8       NaN   
21716   2013      1   25 20:10:00       16:30:00        220       NaN   
24854   2013      1   29 15:59:00       16:05:00         -6       NaN   
25443   2013      1   30 10:25:00       06:35:00        230       NaN   
26522   2013      1   31 14:49:00       14:45:00          4       NaN   
29392   2013     10    3 13:04:00       11:05:00        119       NaN   
36675   2013     10   11 08:39:00       08:32:00          7       NaN   
38272   2013     10   13 08:26:00       08:25:00          1       NaN   
39233   2013     10   14 08:30:00       08:29:00          1       NaN   
40108   2013     10   15 06:57:00       06:25:00         32       NaN   
41895   2013     10   16 20:27:00       20:03:00         24       NaN   
42937   2013     10   17 22:33:00       21:59:00         34       NaN   
51366   2013     10   27 09:46:00       08:05:00        101       NaN   
53256   2013     10   29 08:51:00       09:00:00         -9       NaN   
53486   2013     10   29 13:19:00       13:00:00         19       NaN   
55015   2013     10   31 06:25:00       06:25:00          0       NaN   
55974   2013     11    1 06:54:00       07:00:00         -6       NaN   
56128   2013     11    1 08:55:00       09:05:00        -10       NaN   
56148   2013     11    1 09:12:00       09:15:00         -3       NaN   
56150   2013     11    1 09:13:00       09:21:00         -8       NaN   
...      ...    ...  ...      ...            ...        ...       ...   
319711  2013      9   12 15:10:00       14:50:00         20       NaN   
319716  2013      9   12 15:13:00       14:55:00         18       NaN   
319723  2013      9   12 15:24:00       14:20:00         64       NaN   
319724  2013      9   12 15:25:00       13:16:00        129       NaN   
319725  2013      9   12 15:28:00       14:40:00         48       NaN   
319732  2013      9   12 15:34:00       14:50:00         44       NaN   
319740  2013      9   12 15:43:00       15:28:00         15       NaN   
319748  2013      9   12 15:50:00       14:46:00         64       NaN   
319753  2013      9   12 15:57:00       15:29:00         28       NaN   
319754  2013      9   12 15:57:00       15:20:00         37       NaN   
319755  2013      9   12 15:58:00       16:00:00         -2       NaN   
319756  2013      9   12 15:59:00       15:10:00         49       NaN   
319759  2013      9   12 16:02:00       15:45:00         17       NaN   
319763  2013      9   12 16:13:00       15:30:00         43       NaN   
319766  2013      9   12 16:28:00       16:20:00          8       NaN   
319772  2013      9   12 16:39:00       15:55:00         44       NaN   
319780  2013      9   12 16:57:00       16:59:00         -2       NaN   
319786  2013      9   12 17:06:00       15:15:00        111       NaN   
319797  2013      9   12 17:28:00       16:00:00         88       NaN   
319809  2013      9   12 18:05:00       15:08:00        177       NaN   
319813  2013      9   12 18:15:00       14:23:00        232       NaN   
319912  2013      9   12 21:25:00       18:45:00        160       NaN   
321000  2013      9   13 19:18:00       19:20:00         -2       

* Fill 458 NaN values in arr_time :

In [17]:
fdata['sched_diff'] =pd.DataFrame(fdata["sched_arr_time"]-fdata["sched_dep_time"])
arr_time_new = fdata["dep_time"]+ fdata['sched_diff']
fdata['arr_time'].fillna(arr_time_new,inplace=True)

In [18]:
fdata.isnull().sum()

year                 0
month                0
day                  0
dep_time             0
sched_dep_time       0
dep_delay            0
arr_time             0
sched_arr_time       0
arr_delay         1175
carrier              0
flight               0
tailnum              0
origin               0
dest                 0
air_time          1175
distance             0
hour                 0
minute               0
time_hour            0
sched_diff           0
dtype: int64

* Use TimeDelta for departure and arrival time columns :

In [19]:
fdata["arr_time"]= pd.to_numeric(fdata['arr_time']).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["arr_time"]= pd.to_timedelta(fdata["arr_time"].str[:2]+":"+fdata["arr_time"].str[-2:]+":00")

In [20]:
fdata['arr_delay_na'] =pd.DataFrame(fdata["arr_time"]-fdata["sched_arr_time"])
fdata['arr_delay'].fillna(fdata['arr_delay_na'],inplace=True)

In [21]:
fdata.isnull().sum()

year                 0
month                0
day                  0
dep_time             0
sched_dep_time       0
dep_delay            0
arr_time             0
sched_arr_time       0
arr_delay            0
carrier              0
flight               0
tailnum              0
origin               0
dest                 0
air_time          1175
distance             0
hour                 0
minute               0
time_hour            0
sched_diff           0
arr_delay_na         0
dtype: int64

# ---------------------------------------------------

#  Exploration ideas : 
# 1.  Departure Delays 

In [22]:
len(fdata[fdata.dep_delay>0])

128432

In [23]:
fdata['fdate'] = fdata['year'].astype(str) + "/" + fdata['month'].astype(str) + "/" + fdata['day'].astype(str)

delay_count= pd.DataFrame(fdata[fdata.dep_delay>0].groupby('fdate')[['dep_delay']].count())
delay_count.loc[delay_count['dep_delay'].idxmax()]

dep_delay    674
Name: 2013/12/23, dtype: int64

# 2. Best airport in terms of time departure %

In [24]:
flight_count=pd.DataFrame(fdata.groupby("origin")[["year"]].count())


In [25]:
delay_flights = pd.DataFrame(fdata[fdata.dep_delay>0].groupby('origin')[['dep_delay']].count())

delay_flights.loc[delay_flights['dep_delay'].idxmax()]

dep_delay    52711
Name: EWR, dtype: int64

# 3. Aircraft speed analysis


In [26]:
fdata['flight_speed_Col'] = fdata.distance/(fdata.air_time/60)
flight_speed =pd.DataFrame(fdata.groupby("flight")[["flight_speed_Col"]].count())

In [27]:
flight_speed.loc[flight_speed['flight_speed_Col'].idxmax()]

flight_speed_Col    956
Name: 15, dtype: int64

# 4. On time arrival % analysis

In [28]:
ontime_flights=pd.DataFrame(fdata[fdata.dep_delay<=0].groupby("origin")[["dep_delay"]].count())
ontime_flights.loc[ontime_flights['dep_delay'].idxmax()]

dep_delay    67819
Name: LGA, dtype: int64

# 5. Maximum number of flights headed to some perticular destination

In [29]:
dest_count=pd.DataFrame(fdata.groupby("dest")[["dest"]].count())

dest_count.loc[dest_count['dest'].idxmax()]

dest    16898
Name: ATL, dtype: int64